core/src/main/java/com/microsoft/applicationinsights/internal/statsbeat/AzureMetadataService.java

```diff
@@ -0,0 +1,83 @@
+package com.microsoft.applicationinsights.internal.statsbeat;
+
+import com.microsoft.applicationinsights.internal.channel.common.LazyHttpClient;
+import com.microsoft.applicationinsights.internal.util.ThreadPoolUtils;
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonEncodingException;
+import com.squareup.moshi.Moshi;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.microsoft.applicationinsights.internal.statsbeat.Constants.CUSTOM_DIMENSIONS_OS;
+import static com.microsoft.applicationinsights.internal.statsbeat.Constants.CUSTOM_DIMENSIONS_RP;
+import static com.microsoft.applicationinsights.internal.statsbeat.Constants.RP_VM;
+
+public final class AzureMetadataService {
+
+    private static final Logger logger = LoggerFactory.getLogger(AzureMetadataService.class);
+    private static final AzureMetadataService INSTANCE = new AzureMetadataService();
+    private static final String API_VERSION = "api-version=2017-08-01"; // this version has the smallest payload.
+    private static final String JSON_FORMAT = "format=json";
+    private static final String BASE_URL = "http://169.254.169.254/metadata/instance/compute";
+
+    private static final String endpoint = BASE_URL + "?" + API_VERSION + "&" + JSON_FORMAT;
+    private static final ScheduledExecutorService scheduledExecutor = Executors.newSingleThreadScheduledExecutor(ThreadPoolUtils.createDaemonThreadFactory(AzureMetadataService.class));
+    private JsonAdapter<MetadataInstanceResponse> jsonAdapter;
+
+    public static AzureMetadataService getInstance() {
+        return INSTANCE;
+    }
+
+    public void initialize(long interval) {
+        jsonAdapter = new Moshi.Builder().build().adapter(MetadataInstanceResponse.class);
+        scheduledExecutor.scheduleAtFixedRate(new InvokeMetadataServiceTask(), interval, interval, TimeUnit.SECONDS);
+    }
+
+    public void parseJsonResponse(String response) throws IOException {
+        if (response != null) {
+            MetadataInstanceResponse metadataInstanceResponse = jsonAdapter.fromJson(response);
+            StatsbeatModule.getInstance().getAttachStatsbeat().updateMetadataInstance(metadataInstanceResponse);
+            CustomDimensions.getInstance().getProperties().put(CUSTOM_DIMENSIONS_RP, RP_VM);
+
+            // osType from the Azure Metadata Service has a higher precedence over the running app’s operating system.
+            String osType = metadataInstanceResponse.getOsType();
+            if (osType != null && !"unknown".equalsIgnoreCase(osType)) {
```

Comment:
this will also check the `null` condition:
```suggestion
            if (!"unknown".equalsIgnoreCase(osType)) {
```